### PR TITLE
Fix for 1D operations on 2D residual plots

### LINKED ISSUE #2436: RC1: Slicers and Circular Average fail for residual plots
All slicers and Circular Average 1D plot fail on residual plots.

**To Reproduce**
1. Load a 2D dataset
2. Choose a model
3. Calculate and plot
4. RIght click on the residual chart and choose any slicer or Circular Average
5. Observe the following error
For circular average:
```
09:51:45 - ERROR: Traceback (most recent call last):
  File "sas\qtgui\Plotting\Plotter2D.py", line 308, in onCircularAverage
  File "sas\qtgui\Plotting\Plotter2D.py", line 277, in circularAverage
  File "sas\sascalc\dataloader\manipulations.py", line 605, in __call__
TypeError: 'NoneType' object is not subscriptable
```

for slicers:
```
10:07:36 - ERROR: Traceback (most recent call last):
  File "sas\qtgui\Plotting\Plotter2D.py", line 418, in onBoxAveragingX
  File "sas\qtgui\Plotting\Plotter2D.py", line 358, in setSlicer
  File "sas\qtgui\Plotting\Slicers\BoxSlicer.py", line 481, in __init__
  File "sas\qtgui\Plotting\Slicers\BoxSlicer.py", line 59, in __init__
  File "sas\qtgui\Plotting\Slicers\BoxSlicer.py", line 490, in _post_data
  File "sas\qtgui\Plotting\Slicers\BoxSlicer.py", line 157, in post_data
  File "sas\sascalc\dataloader\manipulations.py", line 460, in __call__
  File "sas\sascalc\dataloader\manipulations.py", line 347, in _avg
TypeError: 'NoneType' object is not subscriptable
```

**SasView version (please complete the following information):**
 - Version: 5.0.6rc1

**Operating system (please complete the following information):**
 - OS: Win10

**Additional context**
This was working in 5.0.5

### PR DESCRIPTION
**Pull request by rozyczko**
## Description

`data.err` can be None for certain plots, like residuals.
This should be checked when data.err is being subscripted.
Note - correct handlers for `None` are present in subsequent code.

Fixes #2436

## How Has This Been Tested?

Tested locally

## Review Checklist (please remove items if they don't apply):

- [x] Code has been reviewed
- [x] Functionality has been tested
- [x] Windows installer (GH artifact) has been tested (installed and worked) 
- [ ] MacOSX installer (GH artifact) has been tested (installed and worked) 
- [x] User documentation is available and complete (if required)
- [x] Developers documentation is available and complete (if required)
- [X] The introduced changes comply with SasView license (BSD 3-Clause)

### CODE REVIEW

File: src/sas/sascalc/dataloader/manipulations.py
Comment (krzywon) [inline src/sas/sascalc/dataloader/manipulations.py:607]: I'm seeing the following error on many of the slicers and think `err_data` needs to be initialized for all manipulations.

> ERROR : sas.qtgui.Perspectives.Fitting.FittingWidget (FittingWidget.py:175) :: Traceback (most recent call last):
  File "sas\qtgui\Plotting\Plotter2D.py", line 380, in onAnnulusView
  File "sas\qtgui\Plotting\Plotter2D.py", line 358, in setSlicer
  File "sas\qtgui\Plotting\Slicers\AnnulusSlicer.py", line 50, in __init__
  File "sas\qtgui\Plotting\Slicers\AnnulusSlicer.py", line 109, in _post_data
  File "sas\sascalc\dataloader\manipulations.py", line 783, in __call__
UnboundLocalError: local variable 'err_data' referenced before assignment

Comment (rozyczko) [reply src/sas/sascalc/dataloader/manipulations.py:607]: of course it should. Doh.
Thanks.

**Comment by krzywon** [comment]
That fixed one issue. One more that I found with sector averaging. The sector calculation attempts to index `err_data` where the traceback fails, so you might need a different check at that line.

Also, all of this work should be ported to `sasdata` once it is approved.

> ERROR: Traceback (most recent call last):
  File "sas\qtgui\Plotting\Plotter2D.py", line 374, in onSectorView
  File "sas\qtgui\Plotting\Plotter2D.py", line 358, in setSlicer
  File "sas\qtgui\Plotting\Slicers\SectorSlicer.py", line 59, in __init__
  File "sas\qtgui\Plotting\Slicers\SectorSlicer.py", line 141, in _post_data
  File "sas\sascalc\dataloader\manipulations.py", line 1026, in __call__
  File "sas\sascalc\dataloader\manipulations.py", line 938, in _agv
TypeError: 'NoneType' object is not subscriptable

**Comment by rozyczko** [comment]
> That fixed one issue. One more that I found with sector averaging. The sector calculation attempts to index `err_data` where the traceback fails, so you might need a different check at that line.
> 
> Also, all of this work should be ported to `sasdata` once it is approved.
> 
> > ERROR: Traceback (most recent call last):
> > File "sas\qtgui\Plotting\Plotter2D.py", line 374, in onSectorView
> > File "sas\qtgui\Plotting\Plotter2D.py", line 358, in setSlicer
> > File "sas\qtgui\Plotting\Slicers\SectorSlicer.py", line 59, in **init**
> > File "sas\qtgui\Plotting\Slicers\SectorSlicer.py", line 141, in _post_data
> > File "sas\sascalc\dataloader\manipulations.py", line 1026, in **call**
> > File "sas\sascalc\dataloader\manipulations.py", line 938, in _agv
> > TypeError: 'NoneType' object is not subscriptable

fixed.

**Review by krzywon — APPROVED** [review]

**Comment by butlerpd** [comment]
do we need an OSX installer test as well for this or are we good to go?

**Comment by butlerpd** [comment]
Also ... this is against 5.0.6. Are we good with that?

**Comment by wpotrzebowski** [comment]
I tested it on Mac and fitting on GPU didn't work for me... It simply didn't update any parameters. Switching to "No OpenCL" fixed the issue. I need to investigate it further if this related issue.

**Comment by rozyczko** [comment]
> Also ... this is against 5.0.6. Are we good with that?

This is a change from 5.0.5 so I'd argue we need to include this fix.

**Comment by butlerpd** [comment]
Is it a new problem and/or does using GPU on windows also fails -- needs a bit more investigation

**Comment by krzywon** [comment]
One thing I just noticed - `installers.yml` is using bumps v0.9.0 which does not have the OpenCL fix, but `release.yml` is using bumps main branch. To properly test this, we need to be sure we are testing the installer that is posted under `Create Release`, not the one under `Installer`. Both yml files use the sasmodels `release_1.0.7` branch.

**Comment by krzywon** [comment]
After more discussion with @butlerpd, he was using the release executable, but the behavior he is seeing is similar to #1596.

**Comment by butlerpd** [comment]
I can now confirm that the behavior I get is the same in the beta. However it is a very subtle bug for me. Start by turning openCL on then load apoferritin data then send to fitting then fit -- all OK. Then turn off openCL change scale parameter to 1 so that I can fit again and error is thrown (and nothing happens on the plots). Try fitting and errors get thrown. Try to change back to OpenCl and only get errors. This was all in one fittab. Now make a new fittab. Same problem fitting throws errors.

However if I start by running a fit with NO OpenCl then turn it on and fit it does work ... and back and forth. I have to say I only tried this last with the beta version but am pretty sure that is the same with the PR so do not see that we should hold up the PR for this (the windows problem anyway).

That said we should maybe revisit #1596 that @krzywon found?

**Comment by wpotrzebowski** [comment]
On Mac, the bug is consistent and it is in this PR but not rc1. After talking to @rozyczko we suspect that this may be due to changes to bumps master branch rather than this PR.

**Comment by wpotrzebowski** [comment]
After some further investigation, we are pretty sure that issue is related to the latest changes to bumps. 

And on Mac is quite easy to check. "No OpenCL" fitting works, OpenCL on GPU fitting doesn't work. I tested  on cylinder model and L-M optimizer.

**Comment by rozyczko** [comment]
The OpenCL issue should be split into a separate bug as it is unrelated to the original report/fix

**Comment by pkienzle** [comment]
Unfortunately the bumps errors are being eaten, but we should be able so show them by changing

https://github.com/SasView/sasview/blob/8ccf562c8d593ababc1d010e3adc632f04037dea/src/sas/sascalc/fit/BumpsFitting.py#L291

to

```
assert …, f"Bumps failed with:\n{result['errors']}"
```

**Pull request merged**
